### PR TITLE
#64 get, create and edit projects

### PR DESCRIPTION
**Pull request by jorgecod**
- Get, Create and Edit Users

### CODE REVIEW

File: src/app/modules/project-management/components/create-project/create-project.component.ts [hunk 1]
```diff
@@ -1,66 +1,61 @@
...
   }
 
-  onSubmit(projectData) {
+  onSubmit(formData) {
+    const projectData = { ...this.projectToEdit, ...formData };
+    if (!this.isEdit) {
+      this.store.dispatch(new actions.PostProject(projectData));
+    } else {
+      this.store.dispatch(new actions.PutProject(projectData));
+    }
```
Comment (enriquezrene) [inline src/app/modules/project-management/components/create-project/create-project.component.ts:53]: In general terms, it's always a good idea to value if(true). In this case, this code should look like:
```
 if (this.isUpdating) {
      this.store.dispatch(new actions.UpdateProject(projectData));
    } else {
      this.store.dispatch(new actions.CreateProject(projectData));
    }
```
I am assuming here the actions have been renamed.

File: src/app/modules/project-management/components/project-list/project-list.component.ts [hunk 2]
```diff
@@ -1,23 +1,36 @@
...
 })
 export class ProjectListComponent implements OnInit {
-  @Input() projects: Project[] = [];
   @Output() editProject = new EventEmitter();
   @Output() deleteProject = new EventEmitter();
 
   projectToDelete: Project;
   openDeleteModal = false;
   filterProjects = '';
+  isLoading: boolean;
```
Comment (enriquezrene) [inline src/app/modules/project-management/components/project-list/project-list.component.ts:19]: I think this flag is not being used so we should not declare this field in here.

File: src/app/modules/project-management/store/project.actions.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import { Action } from '@ngrx/store';
+import { Project } from '../../shared/models';
+
+export enum ProjectActionTypes {
+  GET_PROJECTS = '[Projects] GET_PROJECTS',
+  GET_PROJECTS_SUCCESS = '[Projects] GET_PROJECTS_SUCCESS',
+  GET_PROJECTS_FAIL = '[Projects] GET_PROJECTS_FAIL',
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:7]: Let's replace GET by LOAD

File: src/app/modules/project-management/store/project.actions.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  POST_PROJECT = '[Projects] POST_PROJECT',
+  POST_PROJECT_SUCCESS = '[Projects] POST_PROJECT_SUCCESS',
+  POST_PROJECT_FAIL = '[Projects] POST_PROJECT_FAIL',
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:10]: Let's change POST by CREATE

File: src/app/modules/project-management/store/project.actions.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  PUT_PROJECT = '[Projects] PUT_PROJECT',
+  PUT_PROJECT_SUCCESS = '[Projects] PUT_PROJECT_SUCCESS',
+  PUT_PROJECT_FAIL = '[Projects] PUT_PROJECT_FAIL',
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:13]: Let's change PUT by UPDATE

File: src/app/modules/project-management/store/project.actions.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+export class GetProjectsLoad implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:16]: LoadProjects

File: src/app/modules/project-management/store/project.actions.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.GET_PROJECTS;
+}
+
+export class GetProjectsSuccess implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:20]: LoadProjectsSuccess

File: src/app/modules/project-management/store/project.actions.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  readonly type = ProjectActionTypes.GET_PROJECTS_SUCCESS;
+  constructor(readonly payload: Project[]) {}
+}
+
+export class GetProjectsFail implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:25]: LoadProjectsFail

File: src/app/modules/project-management/store/project.actions.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.GET_PROJECTS_FAIL;
+
+  constructor(public error: string) {}
+}
+
+export class PostProject implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:31]: CreateProject

File: src/app/modules/project-management/store/project.actions.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.POST_PROJECT;
+
+  constructor(public payload: Project) {}
+}
+
+export class PostProjectSuccess implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:37]: CreateProjectSuccess

File: src/app/modules/project-management/store/project.actions.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.POST_PROJECT_SUCCESS;
+
+  constructor(public payload: Project) {}
+}
+
+export class PostProjectFail implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:43]: CreateProjectFail

File: src/app/modules/project-management/store/project.actions.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.POST_PROJECT_FAIL;
+
+  constructor(public error: string) {}
+}
+
+export class PutProject implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:49]: UpdateProject

File: src/app/modules/project-management/store/project.actions.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.PUT_PROJECT;
+
+  constructor(public payload: Project) {}
+}
+
+export class PutProjectSuccess implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:55]: UpdateProjectSuccess

File: src/app/modules/project-management/store/project.actions.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  public readonly type = ProjectActionTypes.PUT_PROJECT_SUCCESS;
+
+  constructor(public payload: Project) {}
+}
+
+export class PutProjectFail implements Action {
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.actions.ts:61]: UpdateProjectFail

File: src/app/modules/project-management/store/project.effects.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        map((project) => {
+          return new actions.GetProjectsSuccess(project);
+        }),
+        catchError((error) => of(new actions.GetProjectsFail(error)))
+      )
+    )
+  );
+
+  @Effect()
+  postProject$: Observable<Action> = this.actions$.pipe(
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.effects.ts:27]: createProject$

File: src/app/modules/project-management/store/project.effects.ts [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        map((projectData) => {
+          return new actions.PostProjectSuccess(projectData);
+        }),
+        catchError((error) => of(new actions.PostProjectFail(error)))
+      )
+    )
+  );
+
+  @Effect()
+  putProject$: Observable<Action> = this.actions$.pipe(
```
Comment (enriquezrene) [inline src/app/modules/project-management/store/project.effects.ts:41]: updateProject$

File: src/app/modules/time-clock/components/project-list-hover/project-list-hover.component.ts [hunk 17]
```diff
@@ -1,27 +1,35 @@
...
   @Output() showFields = new EventEmitter<boolean>();
 
   selectedId: string;
   showButton: number;
   filterProjects = '';
   listProjects: Project[] = [];
+  isLoading: boolean;
 
-  constructor(private projectService: ProjectService) {
+  constructor(private store: Store<any>) {
```
Comment (enriquezrene) [inline src/app/modules/time-clock/components/project-list-hover/project-list-hover.component.ts:21]: Store<any> => Store<ProjectState>

File: src/app/app.module.ts [hunk 18]
```diff
@@ -68,8 +70,11 @@ import { activityManagementReducer } from './modules/activities-management/store
     FormsModule,
     ReactiveFormsModule,
     HttpClientModule,
-    StoreModule.forRoot({ activities: activityManagementReducer }),
-    EffectsModule.forRoot([ActivityEffects]),
+    StoreModule.forRoot(reducers, {
+      metaReducers,
+    }),
+    !environment.production ? StoreDevtoolsModule.instrument() : [],
```
[the remark this answers is not in the document]
Comment (jorgecod) [reply src/app/app.module.ts]: I have updated this line

File: src/app/modules/project-management/components/project-list/project-list.component.ts (hunk 2 above)
Comment (jorgecod) [reply src/app/modules/project-management/components/project-list/project-list.component.ts:19]: yes, we will create an issue to include a loading in the interface, but I think this flag needs to stay here

Comment (enriquezrene) [reply src/app/modules/project-management/components/project-list/project-list.component.ts:19]: gotcha

**Pull request merged**